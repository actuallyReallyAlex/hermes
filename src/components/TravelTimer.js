import React, { useEffect, useState } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Box, Heading } from 'grommet'
import { addCash } from '../redux/actions/user'
import {
  removeCargo,
  setShipLocation,
  setDestination,
  setShipTraveling
} from '../redux/actions/ship'
import moment from 'moment'

const TravelTimer = ({ handleTimerStopped, ship }) => {
  const [differenceInMilliseconds, setDifferenceInMilliseconds] = useState(0)

  // ? Why is this so hard
  // * All you need to do:
  // TODO 1 - ✅ Calculate how long it will take to reach the destination
  // TODO 2 - ✅ Set an ETA in Redux
  // TODO 3 - Set a duration in the TravelTimer based on the ETA in Redux
  // TODO 4 - Start a timer that ticks down the duration every second
  // TODO 5 - Display the time of that duration

  const timerLogic = () => {
    if (ship.isShipTraveling) {
      const travelTimer = setInterval(() => {
        const now = moment()
        const differenceMill = moment(ship.destination.eta, 'x').diff(now) // * gives the number of milliseconds betweeen the eta and now

        const diffDuration = moment.duration({ milliseconds: differenceMill })

        diffDuration.subtract(1, 'second')
        setDifferenceInMilliseconds(diffDuration.asMilliseconds())

        if (diffDuration.milliseconds() === 0) {
          clearInterval(travelTimer)
          handleTimerStopped(ship)
        }
      }, 1000)
    }
  }

  useEffect(timerLogic, [ship.isShipTraveling])

  return ship.isShipTraveling ? (
    <Box>
      <Heading level="3">Travel Timer</Heading>
      <span>{moment(differenceInMilliseconds, 'SSS')}</span>
    </Box>
  ) : null
}

TravelTimer.propTypes = {
  handleTimerStopped: PropTypes.func.isRequired,
  ship: PropTypes.object.isRequired
}

const mapStateToProps = ({ ship }) => ({ ship })

const mapDispatchToProps = dispatch => ({
  handleTimerStopped: ship => {
    const { cargo, destination } = ship
    const sellableItems = cargo.items.filter(
      item => item.destination.value === destination.value
    )
    let profit = 0
    sellableItems.forEach(item => {
      const { quantity, value } = item
      const itemProfit = quantity * value
      profit += itemProfit
    })
    dispatch(addCash(profit))
    sellableItems.forEach(item => {
      dispatch(removeCargo(item))
    })
    dispatch(
      setShipLocation({ name: destination.name, value: destination.value })
    )
    dispatch(setDestination(null))
    dispatch(setShipTraveling(false))
  }
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TravelTimer)
