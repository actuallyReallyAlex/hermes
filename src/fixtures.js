export const defaultState = {
  ship: {
    cargo: {
      items: [],
      volumeRemaining: 5
    },
    location: {
      name: null,
      value: null
    }
  },
  ui: {
    view: 'Ship'
  },
  user: {
    cash: 100
  },
  world: {
    isTimerRunning: false,
    planets: [
      {
        id: "0",
        isHomePlanet: true,
        location: 0,
        items: [
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '0',
            destination: {
              name: 'Test Planet 2',
              value: 50
            }
          },
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '1',
            destination: {
              name: 'Test Planet 2',
              value: 50
            }
          },
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '2',
            destination: {
              name: 'Test Planet 2',
              value: 50
            }
          },
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '3',
            destination: {
              name: 'Test Planet 2',
              value: 50
            }
          },
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '4',
            destination: {
              name: 'Test Planet 2',
              value: 50
            }
          }
        ],
        name: 'Test Planet 1'
      },
      {
        id: "1",
        isHomePlanet: false,
        location: 50,
        items: [
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '5',
            destination: {
              name: 'Test Planet 1',
              value: 0
            }
          },
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '6',
            destination: {
              name: 'Test Planet 1',
              value: 0
            }
          },
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '7',
            destination: {
              name: 'Test Planet 1',
              value: 0
            }
          },
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '8',
            destination: {
              name: 'Test Planet 1',
              value: 0
            }
          },
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '9',
            destination: {
              name: 'Test Planet 1',
              value: 0
            }
          }
        ],
        name: 'Test Planet 2'
      },
      {
        id: "2",
        isHomePlanet: false,
        location: 100,
        items: [
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '10',
            destination: {
              name: 'Test Planet 2',
              value: 50
            }
          },
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '11',
            destination: {
              name: 'Test Planet 2',
              value: 50
            }
          },
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '12',
            destination: {
              name: 'Test Planet 2',
              value: 50
            }
          },
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '13',
            destination: {
              name: 'Test Planet 2',
              value: 50
            }
          },
          {
            name: 'Test Item',
            space: 1,
            value: 1,
            id: '14',
            destination: {
              name: 'Test Planet 2',
              value: 50
            }
          }
        ],
        name: 'Test Planet 3'
      }
    ]
  }
}
