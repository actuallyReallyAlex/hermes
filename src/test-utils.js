import React from 'react'
import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import { defaultState } from './fixtures'
import { render } from '@testing-library/react'

const mockThunks = ({ dispatch, getState }) => next => action => {
  const mockedAction = () => ({ type: 'MOCKED ACTION' })

  let returnValue

  if (typeof action === 'function' || action === undefined) {
    returnValue = mockedAction
  } else {
    returnValue = next(action)
  }

  return returnValue
}

const middlewares = [mockThunks]

export const mockStore = configureStore(middlewares)

export const customRender = options => {
  const { component, state, props } = options

  const Component = component

  return render(
    <Provider store={mockStore(state ? state : defaultState)}>
      <Component {...props} />
    </Provider>
  )
}
