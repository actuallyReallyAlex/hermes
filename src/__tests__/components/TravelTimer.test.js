import { customRender } from '../../test-utils'
import TravelTimer from '../../components/TravelTimer'
import { defaultState } from '../../fixtures'

const customState = {
  ...defaultState,
  ship: {
    ...defaultState.ship,
    destination: {
      name: 'Test Planet 2',
      value: 50
    },
    isShipTraveling: true
  }
}

describe('<TravelTimer />', () => {
  it('Should render the <TravelTimer /> component.', () => {
    const container = customRender({
      component: TravelTimer,
      state: customState
    })
    expect(container.asFragment()).toMatchSnapshot()
  })
})
