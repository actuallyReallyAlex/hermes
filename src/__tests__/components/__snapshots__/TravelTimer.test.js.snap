// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TravelTimer /> Should render the <TravelTimer /> component. 1`] = `
<DocumentFragment>
  <div
    class="StyledBox-sc-13pk1d4-0 IRSNj"
  >
    <h3
      class="StyledHeading-sc-1rdh4aw-0 ftOecI"
    >
      Travel Timer
    </h3>
  </div>
</DocumentFragment>
`;
